include/ginkgo/core/distributed/index_map.hpp

```diff
@@ -0,0 +1,181 @@
+ *   locally owned index space (denoted as index_space::local),
+ * - $r_k: R_k \mapsto [0, \dots, |R_k|)$ which maps global indices into the
+ *   non-locally owned index space (denoted as index_space::non_local).
+ *
+ * The required map can be selected by passing the appropriate type of an
+ * index_space.
+ *
+ * The index map for $I_k$ has no knowledge about any other index maps for
+ * $I_l, l \neq k$. In particular, any global index passed to the `map_to_local`
+ * map that is not part of the specified index space, will be mapped to an
+ * invalid_index.
+ *
+ * \tparam LocalIndexType  type for local indices
+ * \tparam GlobalIndexType  type for global indices
+ */
+template <typename LocalIndexType, typename GlobalIndexType = int64>
+struct index_map {
+    using partition_type = Partition<LocalIndexType, GlobalIndexType>;
+
+    /**
+     * \brief Maps global indices to local indices
+     *
+     * \param global_ids  the global indices to map
+     * \param index_space_v  the index space in which the returned local indices
+     *                       are defined
+     *
+     * \return  the mapped local indices. Any global index that is not in the
+     *          specified index space is mapped to invalid_index.
+     */
+    array<LocalIndexType> map_to_local(
+        const array<GlobalIndexType>& global_ids,
+        index_space index_space_v = index_space::combined) const;
+
+    /**
+     * \brief get size of index_space::local
+     */
+    size_type get_global_size() const;
+
+    /**
+     * \brief get size of index_space::local
+     */
+    size_type get_local_size() const;
+
+    /**
+     * \brief get size of index_space::non_local
+     */
+    size_type get_non_local_size() const;
+
+    /**
+     * \brief Creates a new index map.
+     *
+     * The passed in recv_connections may contain duplicates, which will be
+     * filtered out.
+     *
+     * \param exec  the executor
+     * \param partition  the partition of the global index set
+     * \param rank  the id of the global index space subset
+     * \param recv_connections  the global indices that are not owned by this
+     *                          rank, but accessed by it
+     */
+    index_map(std::shared_ptr<const Executor> exec,
+              std::shared_ptr<const partition_type> partition,
+              comm_index_type rank,
+              const array<GlobalIndexType>& recv_connections);
+
+    /**
+     * \brief Creates an empty index map.
+     */
+    index_map(std::shared_ptr<const Executor> exec);
+
+    /**
+     * \brief get the index set $R_k$ for this rank.
+     *
+     * The indices are ordered by their owning rank and global index.
+     */
+    const segmented_array<GlobalIndexType>& get_remote_global_idxs() const;
+
+    /**
+     * \brief get the index set $R_k$, but mapped to their respective local
+     *        index space.
+     *
+     * The indices are ordered by their owning rank and global index.
```

Comment:
nit: could be a bit more explicit
```suggestion
     * The indices are grouped by their owning rank and sorted according to their global index inside each group.
```